src/snowflake/cli/plugins/connection/util.py

```diff
@@ -50,11 +56,44 @@ def is_regionless_redirect(conn: SnowflakeConnection) -> bool:
         *_, cursor = conn.execute_string(REGIONLESS_QUERY, cursor_class=DictCursor)
         return cursor.fetchone()["REGIONLESS"].lower() == "true"
     except:
-        # by default, assume that
         log.warning("Cannot determine regionless redirect; assuming True.")
         return True
 
 
+def guess_regioned_host_from_allowlist(conn: SnowflakeConnection) -> str | None:
+    """
+    Use SYSTEM$ALLOWLIST to find a regioned host (<account>.x.y.z.snowflakecomputing.com)
+    that corresponds to the given Snowflake connection object.
+    """
+    *_, cursor = conn.execute_string(ALLOWLIST_QUERY, cursor_class=DictCursor)
+    allowlist_tuples = json.loads(cursor.fetchone()["SYSTEM$ALLOWLIST()"])
+    for t in allowlist_tuples:
+        if t["type"] == SNOWFLAKE_DEPLOYMENT:
+            host_parts = t["host"].split(".")
+            if len(host_parts) == 6:
+                return t["host"]
+    return None
+
+
+def get_region(conn: SnowflakeConnection) -> str:
+    """
+    Given a connection, determine the region of the account in the
+    form "x.y.z", or raise an error if we could not find an appropriate host.
+    """
+    if conn.host:
+        host_parts = conn.host.split(".")
+        if len(host_parts) == 6:
+            return ".".join(host_parts[1:4])
+
+    host = guess_regioned_host_from_allowlist(conn)
+    if host:
+        host_parts = host.split(".")
+        if len(host_parts) == 6:
+            return ".".join(host_parts[1:4])
```

Comment:
could encapsulate this into `_get_region_from_host` to avoid duplication.

Reply:
```
region = _get_region_from_host(conn.host) 
    or _get_region_from_host(get_host_from_allowlist(conn))
  
if region is None:
  raise MissingConnectionHostError(host)
return region
  
```

Reply:
Yeah, I messed about with that, but it doesn't really make things any clearer. It saves maybe one line of code, so I decided to keep it expanded like this.